zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/ObserverMaster.java

```diff
@@ -0,0 +1,514 @@
+            return null;
+        }
+        if (pkt.getZxid() != zxid) {
+            final String m = String.format("Unexpected proposal packet on commit ack, expected zxid 0x%d got zxid 0x%d",
+                    zxid, pkt.getZxid());
+            LOG.error(m);
+            throw new RuntimeException(m);
+        }
+        proposedPkts.remove();
+        return pkt;
+    }
+
+    private synchronized void cacheCommittedPacket(final QuorumPacket pkt) {
+        committedPkts.add(pkt);
+        pktsSize += LearnerHandler.packetSize(pkt);
+        // remove 5 packets for every one added as we near the size limit
+        for (int i = 0; pktsSize > pktsSizeLimit * 0.8  && i < 5; i++) {
+            QuorumPacket oldPkt = committedPkts.poll();
+            if (oldPkt == null) {
+                pktsSize = 0;
+                break;
+            }
+            pktsSize -= LearnerHandler.packetSize(oldPkt);
+        }
+        // enforce the size limit as a hard cap
+        while (pktsSize > pktsSizeLimit) {
+            QuorumPacket oldPkt = committedPkts.poll();
+            if (oldPkt == null) {
+                pktsSize = 0;
+                break;
+            }
+            pktsSize -= LearnerHandler.packetSize(oldPkt);
+        }
+    }
+
+    private synchronized void sendPacket(final QuorumPacket pkt) {
+        for (LearnerHandler lh: activeObservers) {
+            lh.queuePacket(pkt);
+        }
+        lastProposedZxid = pkt.getZxid();
+    }
+
+    synchronized void proposalCommitted(long zxid) {
+        QuorumPacket pkt = removeProposedPacket(zxid);
+        if (pkt == null) {
+            return;
+        }
+        cacheCommittedPacket(pkt);
+        sendPacket(pkt);
+    }
+
+    synchronized void informAndActivate(long zxid, long suggestedLeaderId) {
+        QuorumPacket pkt = removeProposedPacket(zxid);
+        if (pkt == null) {
+            return;
+        }
+
+        // Build the INFORMANDACTIVATE packet
+        QuorumPacket informAndActivateQP = Leader.buildInformAndActivePacket(
+                zxid, suggestedLeaderId, pkt.getData());
+        cacheCommittedPacket(informAndActivateQP);
+        sendPacket(informAndActivateQP);
+    }
+
+    synchronized public void start() throws IOException {
+        if (thread != null && thread.isAlive()) {
+            return;
+        }
+        listenerRunning = true;
+        if (self.getQuorumListenOnAllIPs()) {
+            ss = new ServerSocket(port, 10 /* dog science */);
+        } else {
+            ss = new ServerSocket(port, 10 /* dog science */, self.getQuorumAddress().getAddress());
+        }
+        thread = new Thread(this, "ObserverMaster");
+        thread.start();
+        pinger = Executors.newSingleThreadScheduledExecutor();
+        pinger.scheduleAtFixedRate(ping, self.tickTime /2, self.tickTime/2, TimeUnit.MILLISECONDS);
+    }
+
+    public void run() {
+        while (listenerRunning) {
+            try {
+                Socket s = ss.accept();
```

Comment:
One more thing, we added a recent change here to setSoTimeout to initLimit * tickTime to align with what leader is doing.

We should add it here as well.